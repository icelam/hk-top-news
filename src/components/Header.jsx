import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

/* Material UI Components */
import Container from '@material-ui/core/Container';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import Refresh from '@material-ui/icons/Refresh';

/* Components */
import ProgressBar from '@components/ProgressBar';
import SearchInput from '@components/SearchInput';

import logo from '@images/logo-light.svg';

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1
  },
  container: {
    [theme.breakpoints.only('xs')]: {
      paddingRight: theme.spacing(1)
    }
  },
  title: {
    flexGrow: 1,
    [theme.breakpoints.only('xs')]: {
      display: 'none'
    }
  },
  logo: {
    height: theme.spacing(4),
    marginRight: theme.spacing(2)
  },
  search: {
    [theme.breakpoints.only('xs')]: {
      flexGrow: 1
    }
  },
  refreshButton: {
    padding: theme.spacing(1),
    marginLeft: theme.spacing(0.5)
  }
}));

const Header = ({
  onSearch, searchValue, refreshNews, pageLoading
}) => {
  const classes = useStyles();

  return (
    <AppBar className={classes.root}>
      <Container className={classes.container}>
        <Toolbar disableGutters>
          <img id="logo" src={logo} alt="Logo" className={classes.logo} />
          <Typography variant="h6" color="inherit" id="app-name" className={classes.title}>
            香港熱門新聞
          </Typography>

          <SearchInput
            id="search"
            className={classes.search}
            onSearch={onSearch}
            searchValue={searchValue}
          />

          <IconButton
            id="refresh"
            aria-label="refresh"
            aria-controls="menu-appbar"
            aria-haspopup="false"
            color="inherit"
            onClick={refreshNews}
            className={classes.refreshButton}
          >
            <Refresh />
          </IconButton>

        </Toolbar>
      </Container>
      {
        pageLoading ? <ProgressBar /> : null
      }
    </AppBar>
  );
};

Header.propTypes = {
  onSearch: PropTypes.func.isRequired,
  searchValue: PropTypes.string.isRequired,
  refreshNews: PropTypes.func.isRequired,
  pageLoading: PropTypes.bool.isRequired
};

export default Header;
