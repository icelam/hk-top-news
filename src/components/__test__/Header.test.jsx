import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Header from '../Header';

describe('Header', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Header onSearch={() => {}} searchValue="" refreshNews={() => {}} pageLoading={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with app logo, search input box and refresh button', () => {
    const { container } = render(
      <Header onSearch={() => {}} searchValue="" refreshNews={() => {}} pageLoading={false} />
    );
    expect(container.querySelector('#logo')).toBeInTheDocument();
    expect(container.querySelector('#app-name')).toBeInTheDocument();
    expect(container.querySelector('#search')).toBeInTheDocument();
    expect(container.querySelector('#refresh')).toBeInTheDocument();
  });

  it('should show progress bar when  page is loading', () => {
    const { container, rerender } = render(
      <Header onSearch={() => {}} searchValue="" refreshNews={() => {}} pageLoading={false} />
    );
    expect(container.querySelector('.MuiLinearProgress-bar')).toBeNull();

    rerender(
      <Header onSearch={() => {}} searchValue="" refreshNews={() => {}} pageLoading />
    );
    expect(container.querySelector('.MuiLinearProgress-bar')).toBeInTheDocument();
  });

  it('should should trigger onSearch when value in search input changed', () => {
    let query = '';
    const newQuery = '香港';
    const onSearch = jest.fn((event) => {
      query = event.target.value;
    });

    const { container, rerender } = render(
      <Header onSearch={onSearch} searchValue={query} refreshNews={() => {}} pageLoading />
    );

    const searchInput = container.querySelector('#search');
    expect(searchInput.value).toBe('');

    fireEvent.change(searchInput, { target: { value: newQuery } });
    rerender(<Header onSearch={onSearch} searchValue={query} refreshNews={() => {}} pageLoading />);
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(searchInput.value).toEqual(newQuery);
  });

  it('should should trigger refreshNews when refresh button is clicked', () => {
    const refreshNews = jest.fn();
    const { container } = render(
      <Header onSearch={() => {}} searchValue="" refreshNews={refreshNews} pageLoading={false} />
    );

    const refreshButton = container.querySelector('#refresh');
    fireEvent.click(refreshButton);
    expect(refreshNews).toHaveBeenCalledTimes(1);
  });
});
