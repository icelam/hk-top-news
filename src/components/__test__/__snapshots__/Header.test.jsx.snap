// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-root-1 mui-fixed MuiPaper-elevation4"
  >
    <div
      class="MuiContainer-root makeStyles-container-2 MuiContainer-maxWidthLg"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular"
      >
        <img
          alt="Logo"
          class="makeStyles-logo-4"
          id="logo"
          src="logo-light.svg"
        />
        <h6
          class="MuiTypography-root makeStyles-title-3 MuiTypography-h6 MuiTypography-colorInherit"
          id="app-name"
        >
          香港熱門新聞
        </h6>
        <div
          class="makeStyles-search-7 makeStyles-search-5"
        >
          <div
            class="makeStyles-searchIcon-8"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <div
            class="MuiInputBase-root makeStyles-inputRoot-9"
          >
            <input
              aria-label="search"
              class="MuiInputBase-input makeStyles-inputInput-10"
              id="search"
              placeholder="搜尋"
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          aria-controls="menu-appbar"
          aria-haspopup="false"
          aria-label="refresh"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-refreshButton-6 MuiIconButton-colorInherit"
          id="refresh"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
